snapcraft/plugins/gomod.py

```diff
@@ -0,0 +1,223 @@
+      can be set to "off" to disable all module authentication. Default is
+      "sum.golang.org".
+
+    - go-no-sumdb:
+      (list of strings)
+      List of module patterns that will not be authenticated with the
+      module checksum database. Each value is a glob pattern that is
+      prefix matched to the required modules. Default is to authenticate
+      all modules.
+
+    - go-private:
+      (list of strings)
+      List of module patterns that are private, these modules will not
+      be downloaded from a proxy, or authenticated against the checksum
+      database. Each value is a glob pattern that is prefix matched to
+      the required modules. Default is that there are no private modules.
+
+"""
+
+
+import os
+
+import snapcraft
+
+
+class GoModPlugin(snapcraft.BasePlugin):
+    @classmethod
+    def schema(cls):
+        schema = super().schema()
+        schema["properties"]["go-buildtags"] = {
+            "type": "array",
+            "minitems": 1,
+            "uniqueItems": True,
+            "items": {"type": "string"},
+            "default": [],
+        }
+        schema["properties"]["go-channel"] = {
+            "type": "string",
+            "default": "latest/stable",
+        }
+        schema["properties"]["go-no-proxy"] = {
+            "type": "array",
+            "minitems": 1,
+            "uniqueItems": True,
+            "items": {"type": "string"},
+            "default": [],
+        }
+        schema["properties"]["go-no-sumdb"] = {
+            "type": "array",
+            "minitems": 1,
+            "uniqueItems": True,
+            "items": {"type": "string"},
+            "default": [],
+        }
+        schema["properties"]["go-packages"] = {
+            "type": "array",
+            "minitems": 1,
+            "uniqueItems": True,
+            "items": {"type": "string"},
+        }
+        schema["properties"]["go-private"] = {
+            "type": "array",
+            "minitems": 1,
+            "uniqueItems": True,
+            "items": {"type": "string"},
+            "default": [],
+        }
+        schema["properties"]["go-proxy"] = {
+            "type": "array",
+            "minitems": 1,
+            "uniqueItems": True,
+            "items": {"type": "string"},
+            "default": ["https://proxy.golang.org", "direct"],
+        }
+        schema["properties"]["go-sumdb"] = {
+            "type": "string",
+            "default": "sum.golang.org",
+        }
+        schema["required"] = ["go-packages"]
+        return schema
+
+    @classmethod
+    def get_build_properties(cls):
+        return ["go-packages", "go-buildtags"]
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+
+        self._env = [
+            "GO111MODULE=on",
+            "GOPROXY={}".format(",".join(options.go_proxy)),
+            "GOSUMDB={}".format(options.go_sumdb),
+        ]
+        if options.go_no_proxy:
```

Comment:
These three options seem to be doable from just using snapcraft's `build-environment` keyword. Same for the three above.